### PR TITLE
feat(cloudfront): Key Value Store L2

### LINKED ISSUE #28377: aws-cloudfront: support KeyValueStore
### Describe the feature

Hi! Are there any plans of supporting the new Cloudfront [KeyValueStore](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/kvs-with-functions.html) feature in cdk?

### Use Case

I'd like to
1. create a key value store, and
2. associate the store with a function.

### Proposed Solution

_No response_

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.115.0

### Environment details (OS name and version, etc.)

Ubuntu 22

### PR DESCRIPTION
**Pull request by laurelmay**
This adds an initial resource to support creating a Key Value Store and
specifying an import source. Unfortunately, CloudFormation doesn't seem
to support specifying the `KeyValueStoreAssociations` property of a
function so there isn't a way to actually associate the store with a
function.

Closes #28377.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-cloudfront/lib/key-value-store.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  readonly keyValueStoreStatus: string;
+}
+
+/**
+ * A CloudFront Key Value Store.
+ *
+ * @resource AWS::CloudFront::KeyValueStore
+ */
+export class KeyValueStore extends Resource implements IKeyValueStore {
+
```
Comment (kaizencc) [inline packages/aws-cdk-lib/aws-cloudfront/lib/key-value-store.ts:165]:
```suggestion
```

File: packages/aws-cdk-lib/aws-cloudfront/test/key-value-store.test.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
+import * as path from 'node:path';
+import { Match, Template } from '../../assertions';
+import * as s3 from '../../aws-s3';
+import { App, Stack } from '../../core';
+import { KeyValueStore, ImportSource } from '../lib';
+
+describe('Key Value Store', () => {
+
```
Comment (kaizencc) [inline packages/aws-cdk-lib/aws-cloudfront/test/key-value-store.test.ts:8]:
```suggestion
```

File: packages/aws-cdk-lib/aws-cloudfront/test/key-value-store.test.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    // GIVEN
+    const stack = new Stack();
+
+    // WHEN
+    const store = KeyValueStore.fromKeyValueStoreArn(stack, 'TestStore', 'arn:aws:cloudfront::123456789012:key-value-store/id1');
+
+    // THEN
+    expect(() => store.keyValueStoreStatus).toThrow('Status is not available for imported Key Value Store');
+  });
+
```
Comment (kaizencc) [inline packages/aws-cdk-lib/aws-cloudfront/test/key-value-store.test.ts:123]:
```suggestion
```

File: packages/aws-cdk-lib/aws-cloudfront/lib/key-value-store.ts [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+   * Import a Key Value Store using its ARN.
+   */
+  public static fromKeyValueStoreArn(scope: Construct, id: string, keyValueStoreArn: string): IKeyValueStore {
+    const storeId = Arn.split(keyValueStoreArn, ArnFormat.SLASH_RESOURCE_NAME).resourceName;
+    if (!storeId) {
+      throw new Error(`Invalid Key Value Store ID '${keyValueStoreArn}'`);
+    }
+    return new class Import extends Resource implements IKeyValueStore {
+      readonly keyValueStoreArn: string = keyValueStoreArn;
+      readonly keyValueStoreId: string = storeId!;
```
Comment (kaizencc) [inline packages/aws-cdk-lib/aws-cloudfront/lib/key-value-store.ts:176]: Ugh i feel like TS should know that `storeId` exists here. Does it not?

**Review by kaizencc — COMMENTED** [review]
This looks great @kylelaker

File: packages/aws-cdk-lib/aws-cloudfront/lib/key-value-store.ts (hunk 4 above)
Comment (kaizencc) [reply packages/aws-cdk-lib/aws-cloudfront/lib/key-value-store.ts:175]: That's dumb. It does not. Oh well

**Review by kaizencc — APPROVED** [review]

**Review by kaizencc — DISMISSED** [review]

File: packages/aws-cdk-lib/aws-cloudfront/lib/key-value-store.ts [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+ * @resource AWS::CloudFront::KeyValueStore
+ */
+export class KeyValueStore extends Resource implements IKeyValueStore {
+  /**
+   * Import a Key Value Store using its ARN.
+   */
+  public static fromKeyValueStoreArn(scope: Construct, id: string, keyValueStoreArn: string): IKeyValueStore {
+    const storeId = Arn.split(keyValueStoreArn, ArnFormat.SLASH_RESOURCE_NAME).resourceName;
+    if (!storeId) {
+      throw new Error(`Invalid Key Value Store ID '${keyValueStoreArn}'`);
```
Comment (kylelaker) [inline packages/aws-cdk-lib/aws-cloudfront/lib/key-value-store.ts:171]:
```suggestion
      throw new Error(`Invalid Key Store Arn: '${keyValueStoreArn}'`);
```

File: packages/aws-cdk-lib/aws-cloudfront/test/key-value-store.test.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  test('imported resource throws error when missing ID', () => {
+    // GIVEN
+    const stack = new Stack();
+
+    // THEN
+    expect(
+      () => KeyValueStore.fromKeyValueStoreArn(stack, 'TestStore', 'arn:aws:cloudfront::123456789012:key-value-store'),
+    ).toThrow(/Invalid Key Store Arn:/);
```
Comment (kylelaker) [inline packages/aws-cdk-lib/aws-cloudfront/test/key-value-store.test.ts:109]:
```suggestion
    ).toThrow(/Invalid Key Value Store Arn:/);
```

File: packages/aws-cdk-lib/aws-cloudfront/README.md [hunk 7]
```diff
@@ -536,6 +536,27 @@ It will auto-generate the name of the function and deploy it to the `live` stage
...
+```ts
+const store = new cloudfront.KeyValueStore(this, 'KeyValueStore');
+```
+
+To also include an initial set of value, the `source` property can be specified. For the
+structure of this file, see [Creating a file of key value pairs](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/kvs-with-functions-create-s3-kvp.html).
+
+```ts
+const store = new cloudfront.KeyValueStore(this, 'KeyValueStore', {
+  name: 'KeyValueStore',
```
Comment (kylelaker) [inline packages/aws-cdk-lib/aws-cloudfront/README.md:555]:
```suggestion
  keyValueStoreName: 'KeyValueStore',
```

**Review by kaizencc — APPROVED** [review]

**Comment by kylelaker** [comment]
@kaizencc It looks like Mergify doesn't want to update because of #28481 modifying a file in `.github/workflows`. I am going to go ahead and manually update this PR which will probably dismiss your review.

File: packages/aws-cdk-lib/aws-cloudfront/lib/key-value-store.ts (hunk 4 above)
Comment (kylelaker) [reply packages/aws-cdk-lib/aws-cloudfront/lib/key-value-store.ts:175]: Yeah; it looks like TS realizes the type is `string` before the class definition but within it, it sees it as `string | undefined`

**Review by kaizencc — APPROVED** [review]

**Pull request merged**